appdata/system/man/yotta.1
--[[pod_format="raw",created="2024-03-18 18:29:39",modified="2024-03-18 19:15:40",revision=14]]
NAME
\tyotta -- dependency and package manager

SYNOPSIS
\tyotta <commands>

COMMANDS
\tinit
\t\tinitialize a yottafile for the current directory
\tlist
\t\tlist the current dependencies
\t\t-v - verbose: list all tracked files under each dependency
\tadd <cart>
\t\tadd a cart ref as dependency (but don't add/remove anything yet)
\tremove <cart>
\t\tremove a cart ref as dependency (but don't add/remove anythin yet)
\tapply
\t\tmake the current directory match a yottafile
\t\t(this is what actually adds and removes files!)
\tforce
\t\tthis is an apply that will reinstall every dep (for updates?)
\tutil install <cart>
\t\tglobally install utility cart ref
\tutil uninstall <cart>
\t\tglobally uninstall utility cart ref
\tutil list
\t\tlist all tracked util cart refs and their files

DESCRIPTION
\tThe yotta command seems complicated but it's pretty simple if you've ever used dependency managers or package managers on other systems or for other programming languages. I'll outline its two major functions separately.

PACKAGE MANAGEMENT
\t1.\t\^i> yotta util install /my_utils/cool_util.p64 #bbs_system_package\^-i
\t2.\tDone!
\t3.\tWait, I want one of them gone, it sucks!
\t4.\t\^i> yotta util uninstall #bbs_system_package\^-i
\t5.\tDone!

Let's say you see a cool yotta-compatible system package like a new command-line utility (like 'new' or 'merge' ;) ) on the BBS, or let's say you've written a yotta-compatible cartridge containing your utility. Or both!

Grab your terminal, its location doesn't matter. If it's on the BBS, the package cartridge has a BBS ID, and you can supply that like you would with the 'load' command: 

\t\^i> yotta util install #bbs_system_package\^-i

If your dependency lives in a file on your Picotron, you can specify the path:

\t\^i> yotta util install /my_utils/cool_util.p64\^-i

This will track the dependency by adding it to the system yottafile at /appdata/system/global_yottafile.pod and will immediately install it. You can specify multiple packages at once from mixed sources, also, like:

\t\^i> yotta util install /my_utils/cool_util.p64 #bbs_system_package\^-i

If you decide you no longer want or need a package, you can remove it nearly the same way you add it:

\t\^i> yotta util uninstall /my_utils/cool_util.p64\^-i

It will be immediately uninstalled, but it won't remove any files created by the utility or whatever, only what was installed originally. If it leaves empty directories after it's done removing files, it will clean those up too.

DEPENDENCY MANAGEMENT
\t1.\t\^i> cd /ram/cart\^-i
\t2.\t\^i> yotta init\^-i
\t3.\t\^i> yotta add #bbs_id /my_libs/library.p64\^-i
\t4.\t\^i> yotta apply\^-i
\t5.\tIn your lua, include "./lib/bbs_id/main.lua" and include "./lib/library/main.lua"
\t6.\tDone!

Let's say you've got your cartridge, and you'd like to use a library that exists on the BBS that is compatible with yotta. Or... let's say you have some yotta-compatible libraries on your system that you've written or downloaded and you'd like to use their features in your current cartridge. Or both!

Sit your terminal in /ram/cart and prepare the cartridge for use with yotta:

\t\^i> yotta init\^-i

This will create a /ram/cart/lib directory, and a /ram/cart/lib/yottafile.pod file to track dependancies and what dependancies own what files.

Now, to add a dependency, you need to know where it lives. If it's on the BBS, the library cartridge has a BBS ID, and you can supply that like you would with the 'load' command:

\t\^i> yotta add #bbs_id\^-i

If your dependency lives in a file on your Picotron, you can specify the path:

\t\^i> yotta add /my_libs/library.p64\^-i

This will track the dependency by adding it to the yottafile, but it won't immediately install it. You can specify multiple packages at once from mixed sources, like:

\t\^i> yotta add #bbs_id /my_libs/library.p64\^-i

When you're ready to install the dependencies into your cart's lib/ folder, run:

\t\^i> yotta apply.\^-i

It will fetch the BBS carts and copy the filesystem carts and unpack each of their exports/ folders into named directories in libs/. For example, using the above two sources we added, you could expect to see something in your cart like libs/bbs_id/main.lua and libs/library/main.lua. Everything present in the target dependency cart's exports folder will be copied into this lib/[name] directory (except for .map, .sfx, and .gfx files).

If a target cartridge's export folder contains .map, .sfx, or .gfx files, they will be copied into your cartridge's map/, sfx/, and gfx/ directories, respectively, named [library_name]_[exported_filename].[ext]. They will still be tracked by yotta, so if you remove the dependency and do a yotta apply, it will remove them cleanly.

If your dependency targets update somehow, you can use yotta force to force it to reacquire all dependencies instead of simply passing over already-installed packages.

If you decide you no longer need a package, you can remove it nearly the same way you add it:

\t\^i> yotta remove #bbs_id\^-i

Then, the next yotta apply will identify this and remove lib/bbs_id from your cart, and delete any sfx/gfx/map files that were added by it. It won't remove the includes from your lua code - that's up to you!

MAKING YOTTA-COMPATIBLE LIBRARIES
It's simple! Place whatever you'd like to be available in your end-user cart's lib/your_package_name directory into your cart's exports/ folder. Make it if it doesn't exist.
If you add .sfx, .map, or .gfx files, they will be copied into your end-users sfx/, map/, and gfx/ folders, respectively, named [your_package_name]_[filename_in_your_exports].[ext]

If your cartridge does nothing else other than offer your library to folks, then you're welcome to use my library container utility (#lib_utilinst) to help inform users that this is not a 'runnable' cartridge and is instead a kind of 'resource' cartridge. It gives them the ability to view the files if they run it, but not much else.

Once you publish your cart with the exports/ folder to the BBS, other people using yotta will be able to use it in their projects with:

\t\^i> yotta add #your_bbs_cart_id\^-i

\tfollowed by a:

\t\^i> yotta apply\^-i

I propose a convention of providing a primary 'include entry point' (if applicable) of 'main.lua', so placing your library file in your /ram/cart/exports/main.lua. But, all files in your exports/ folder will end up in the end-user's lib/your_package_name/ directory, so you can do as you will.

MAKING YOTTA-COMPATIBLE PACKAGES
It's also simple! Make an exports folder in your cartridge at /ram/cart/exports. Assume the contents of your cartridge's exports directory will be overlaid atop the end-user's Picotron drive starting at the root level. So, if you'd like to package a system-wide terminal utility, you would want to ensure that /ram/cart/exports/appdata/system/util exists as a folder, and inside of that, you would place your utility (like new.lua!).

When end users use yotta to install your package cartridge, it will copy the exports tree over the root filesystem, and when users instruct yotta to uninstall it, it will remove the files present in your export tree from the user's filesystem (and will clean up empty directories if it leaves them).

If you don't want to require your end users to use yotta to install your package, and your cart does nothing else but hold the package, you could use my package installer (#lib_utilinst, same as my library container). It will let people either manually copy the files out of /ram/cart/exports or will offer to install/uninstall it to their with similar logic to yotta (but without requiring yotta to be installed) by pressing X while running it.

AUTHOR
\tCreated by ahrotahn <reh@ahrotahn.net>


appdata/system/util/yotta.lua
--[[pod_format="raw",created="2024-03-16 21:44:35",modified="2024-03-21 13:18:39",revision=216]]
-- yotta: a "package" "manager" for Picotron
--  by ahrotahn <reh@ahrotahn.net>
-- v1.0: initial hideous beautiful version

YF_VER = 1
YF_DIR = "lib"
YF_TRACKFILE = "yottafile.pod"
YF_DEFAULT = YF_DIR.."/" .. YF_TRACKFILE
YF_SYS = "/appdata/system/global_yottafile.pod"

function usage()
	print("usage: yotta [command]")
	print("commands:")
	print(" init - initialize a yottafile for the current directory")
	print(" list - list the current dependencies")
	print("   -v - verbose: list all tracked files under each dependency")
	print(" add [cart] - add a cart ref as dependency (but don't add/remove anything yet)")
	print(" remove [cart] - remove a cart ref as dependency (but don't add/remove anythin yet)")
	print(" apply - make the current directory match a yottafile")
	print("   (this is what actually adds and removes files!)")
	print(" force - this is an apply that will reinstall every dep (for updates?)")
	print(" util install [cart] - globally install utility cart ref")
	print(" util uninstall [cart] - globally uninstall utility cart ref")
	print(" util list - list all tracked util cart refs and their files")
end

function download_bbs_cart(id)
	-- ok, yes, yes zep, i know you said this isn't a public api
	-- but i do so like to live dangerously.
	-- return fetch("http://picotron-dev.local/"..id)
	return fetch("https://www.lexaloffle.com/bbs/get_cart.php?cat=8&lid="..id)
	-- i'm placing it up top in an easy-to-access place in case
	-- it needs to be manually edited by end-users in the future
	-- to comply with some kind of eventual future public api
	-- to get everything running again.
end

function read_yottafile(yfp)
	yfp = yfp or YF_DEFAULT
	local raw = fetch(yfp)
	
	if raw["_yf_version"] == 1 then
		return {
			tags = raw["tags"],
			deps = raw["deps"],
			tracked = raw["tracked"]
		}
	else
		print("ERROR: unknown yottafile version: " .. raw["_yf_version"])
		print("Malformed or future yottafile? Maybe update yotta or pick a real target?")
		exit(1)
	end
end

function write_yottafile(yf, yfp)
	yfp = yfp or YF_DEFAULT
	store(yfp, {
		["_yf_version"] = YF_VER,
		tags = yf["tags"],
		deps = yf["deps"],
		tracked = yf["tracked"]
	})
end

function normalize_cart_ref(ref)
	if ref:sub(1,1) == "#" then
		-- BBS id
		ref = "_bbs_" .. ref:sub(2)
	else
		-- filename
		ref = fullpath(ref):basename()
		if (ref:sub(-4) == ".png") ref = ref:sub(1, -5)
		if (ref:sub(-4) == ".p64") ref = ref:sub(1, -5)
	end
	
	ref = table.concat(split(ref, "-", false), "_")
	ref = table.concat(split(ref, ".", false), "_")
	-- TODO: figure out more bad fs symbols to fix probably lmao

	return ref
end

function install_ref(ref, yfp, dest_dir, owner_ref)
	yfp = yfp or YF_DEFAULT
	owner_ref = owner_ref or ref
	
	local norm_name = normalize_cart_ref(ref)
	dest_dir = dest_dir or "./"..YF_DIR.."/"..norm_name
	local tmp_cart = "/ram/_yotta_temp.p64"

	-- fetch cart data
	if ref:sub(1,1) == "#" then
		-- BBS download
		tmp_cart ..= ".png"
		local data = download_bbs_cart(ref:sub(2))	
			
		if type(data) != "string" or #data <= 0 then
			print("ERROR: could not acquire BBS cart with ID: " .. ref);
			exit(1)
		end
		
		rm(tmp_cart)
		store(tmp_cart, data)
	else
		-- file path
		
		local filename = fullpath(ref)
		if fstat(filename) != "folder" then
			filename = filename..".p64"
			if fstat(filename) != "folder" then
				filename = filename..".png"
				if fstat(filename) != "folder" then
					print("ERROR: could not acquire filesystem cart with path: " .. ref)
					exit(1)
				end
			end
		end

		if (filename:sub(-8) == ".p64.png") tmp_cart ..= ".png"
		cp(filename, tmp_cart)
	end
	
	-- does target have /export
	local exports_dir = tmp_cart .. "/exports"
	if not fstat(exports_dir) then
		print("WARNING: specified ref ("..ref..") has no exports. IGNORING REF!")
		return false
	end
	
	if yfp == YF_SYS then
		-- overlay exports onto root fs
		-- kind of yikes ig
		local result = merge(exports_dir, "")
		foreach(result["files"], function(v) yf_add_track(ref, v, yfp) end)
		foreach(result["dirs"], function(v) yf_add_track(ref, v, yfp) end)
	else
		if fstat(dest_dir) != "folder" then
			mkdir(dest_dir)
			yf_add_track(ref, dest_dir, yfp)
		end
	
		-- extract exports into appropriate CWD locations
		for exp in all(ls(exports_dir)) do
			local full = fullpath(exports_dir.."/"..exp)
			local ext = exp:sub(-4)
			local dest = dest_dir.."/"..full:basename()
			local lua = (ext == ".lua")
		
			if ext == ".gfx" or ext == ".sfx" or ext == ".map" then
				-- allow gfx, sfx, and map exports!
				dest = "./"..ext:sub(-3).."/"..norm_name.."_"..full:basename()
			end
		
			print("copying src:"..exp.." => dst:"..dest)
			cp(full, fullpath(dest))
			yf_add_track(ref, dest, yfp)
		end
	end

	-- clean up...
	rm(tmp_cart)
end

function yf_add_track(ref, file, yfp)
	local yf = read_yottafile(yfp)
	local norm = normalize_cart_ref(ref)
	if type(yf["tracked"][norm]) != "table" then
		yf["tracked"][norm] = {}
	end
	if count(yf["tracked"][norm], file) == 0 then
		add(yf["tracked"][norm], file)
	end
	if type(yf["tracked"]["_refs"]) != "table" then
		yf["tracked"]["_refs"] = {}
	end
	if count(yf["tracked"]["_refs"], ref) == 0 then
		add(yf["tracked"]["_refs"], ref)
	end
	write_yottafile(yf, yfp)
end

function yf_remove_track(ref, file, yfp)
	local yf = read_yottafile(yfp)
	local norm = normalize_cart_ref(ref)
	if type(yf["tracked"][norm]) != "table" then
		return
	end
	while count(yf["tracked"][norm], file) > 0 do
		del(yf["tracked"][norm], file)
	end
	if #yf["tracked"][norm] == 0 then
		deli(yf["tracked"], norm)
		del(yf["tracked"]["_refs"], ref)
	end
	write_yottafile(yf, yfp)
end

function uninstall_ref(ref, yfp)
	yfp = yfp or YF_DEFAULT
	local norm_name = normalize_cart_ref(ref)
	local yf = read_yottafile(yfp)
	
	if type(yf["tracked"][norm_name]) != "table" then
		return
	end
	
	for file in all(yf["tracked"][norm_name]) do
		local type = fstat(fullpath(file))
		if type != "folder" then
			rm(fullpath(file))
			yf_remove_track(ref, file, yfp)
		end
	end
	
	for file in all(yf["tracked"][norm_name]) do
		if type == "folder" then
			if #ls(fullpath(file)) == 0 then
				-- remove folders originally added by this ref that are now empty
				rm(fullpath(file))
			end
			
			yf_remove_track(ref, file, yfp)
		end
	end
end

function merge(src, dst, dir)
	dir = dir or "/"
	
	local dirs = {}
	local files = {}
	
	for entry in all(ls(src..dir)) do
		local type = fstat(src..dir..entry)
		local rpath = dir..entry
		if type == "folder" then
			if fstat(dst..rpath) != "folder" then
				mkdir(dst..rpath)
			end

			local inner = merge(src, dst, dir..entry.."/")
			foreach(inner["dirs"], function(v) add(dirs, v) end)
			foreach(inner["files"], function(v) add(files, v) end)
		elseif type == "file" then
			if (fstat(dst..rpath) == "file") then
				cp(dst..rpath, dst..rpath..".bak")
			end
			
			add(files, dst..rpath)
			cp(src..rpath, dst..rpath)
		end
	end
	
	return {
		dirs = dirs,
		files = files
	}
end

function yf_init(yfp)
	yfp = yfp or YF_DEFAULT
	
	if fstat(yfp) then
		return
	end

	local tags = {
		["created"] = date()
	}
	
	print("initializing yottafile at: " .. fullpath(yfp))
	
	if yfp != YF_SYS and fstat(YF_DIR) != "folder" then
		mkdir(YF_DIR)
	end
	
	write_yottafile({
		["deps"] = {},
		["tags"] = tags,
		["tracked"] = {
			["_refs"] = {}
		}
	}, yfp)
end

function yf_list(verbose, yfp)
	verbose = verbose or false
	yfp = yfp or YF_DEFAULT
	
	local yf = read_yottafile(yfp)
	print("list of current "..(yfp == YF_SYS and "system-level utilities" or "dependencies")..":")
	
	for dep in all(yf["deps"]) do
		local norm = normalize_cart_ref(dep)
		print("  "..dep.." (\""..norm.."\")")
		if verbose or util then
			for file in all(yf["tracked"][norm]) do
				print("    " .. file)
			end
		end
	end
end	

--
-- entry
--

cd(env().path)
local argv = env().argv
if (#argv < 1) then
	usage()
	exit(1)
end

local command = argv[1]

if count({"init", "util", "help"}, command) == 0
and fstat(YF_DEFAULT) != "file" then
	print("ERROR: No yottafile present")
	print("")
	usage()
	exit(1)
end

if command == "init" then
	yf_init()
elseif command == "list" then
	yf_list(argv[2] == "-v")
elseif command == "add" then
	for i = 2, #argv do
		local ref = argv[i]
		print("adding "..ref)

		local yf = read_yottafile()
		add(yf["deps"], ref)
		write_yottafile(yf)
	end
elseif command == "remove" then
	for i = 2, #argv do
		local ref = argv[i]
		print("removing "..ref)
	
		local yf = read_yottafile()
		del(yf["deps"], ref)
		write_yottafile(yf)
	end
elseif command == "apply" or command == "force" then
	local yf = read_yottafile()
	local force = (command == "force")

	for ref in all(yf["tracked"]["_refs"]) do
		if count(yf["deps"], ref) == 0 then
			print("removing dependency: "..ref)
			uninstall_ref(ref)
		end
	end
	
	for ref in all(yf["deps"]) do
		if force or count(yf["tracked"]["_refs"]) == 0 then
			print((force and "updating" or "adding") .. " dependency: "..ref)
			install_ref(ref)
		end
	end
elseif command == "util" then
	yf_init(YF_SYS)
	
	subcommand = argv[2]
	if subcommand == "install" then
		for i = 3, #argv do
			local ref = argv[i]
			print("installing utility: "..ref)

			local yf = read_yottafile(YF_SYS)
			add(yf["deps"], ref)
			write_yottafile(yf, YF_SYS)
		
			install_ref(ref, YF_SYS)
		end
	elseif subcommand == "uninstall" then
		for i = 3, #argv do
			local ref = argv[i]
			print("uninstalling utility: "..ref)
		
			local yf = read_yottafile(YF_SYS)
			del(yf["deps"], ref)
			write_yottafile(yf, YF_SYS)
				
			uninstall_ref(ref, YF_SYS)
		end
	elseif subcommand == "list" then
		yf_list(true, YF_SYS)
	else
		usage()
		exit((subcommand != "help") and 1 or 0)
	end
else
	usage()
	exit((command != "help") and 1 or 0)
end